package/gluon-mesh-batman-adv-core/src/respondd.c

```diff
@@ -0,0 +1,606 @@
+*/
+
+
+#include <respondd.h>
+
+#include <iwinfo.h>
+#include <json-c/json.h>
+#include <libgluonutil.h>
+
+#include <alloca.h>
+#include <glob.h>
+#include <inttypes.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <arpa/inet.h>
+#include <net/if.h>
+#include <netinet/in.h>
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+
+#include <linux/ethtool.h>
+#include <linux/if_addr.h>
+#include <linux/sockios.h>
+
+
+
+#define _STRINGIFY(s) #s
+#define STRINGIFY(s) _STRINGIFY(s)
+
+
+static struct json_object * get_addresses(void) {
+        FILE *f = fopen("/proc/net/if_inet6", "r");
+        if (!f)
+                return NULL;
+
+        char *line = NULL;
+        size_t len = 0;
+
+        struct json_object *ret = json_object_new_array();
+
+        while (getline(&line, &len, f) >= 0) {
+                /* IF_NAMESIZE would be enough, but adding 1 here is simpler than subtracting 1 in the format string */
+                char ifname[IF_NAMESIZE+1];
+                unsigned int flags;
+                struct in6_addr addr;
+                char buf[INET6_ADDRSTRLEN];
+
+                if (sscanf(line,
+                           "%2"SCNx8"%2"SCNx8"%2"SCNx8"%2"SCNx8"%2"SCNx8"%2"SCNx8"%2"SCNx8"%2"SCNx8
+                           "%2"SCNx8"%2"SCNx8"%2"SCNx8"%2"SCNx8"%2"SCNx8"%2"SCNx8"%2"SCNx8"%2"SCNx8
+                           "  %*2x %*2x %*2x %2x %"STRINGIFY(IF_NAMESIZE)"s",
+                           &addr.s6_addr[0], &addr.s6_addr[1], &addr.s6_addr[2], &addr.s6_addr[3],
+                           &addr.s6_addr[4], &addr.s6_addr[5], &addr.s6_addr[6], &addr.s6_addr[7],
+                           &addr.s6_addr[8], &addr.s6_addr[9], &addr.s6_addr[10], &addr.s6_addr[11],
+                           &addr.s6_addr[12], &addr.s6_addr[13], &addr.s6_addr[14], &addr.s6_addr[15],
+                           &flags, ifname) != 18)
+                        continue;
+
+                if (strcmp(ifname, "br-client"))
+                        continue;
+
+                if (flags & (IFA_F_TENTATIVE|IFA_F_DEPRECATED))
+                        continue;
+
+                inet_ntop(AF_INET6, &addr, buf, sizeof(buf));
+
+                json_object_array_add(ret, json_object_new_string(buf));
+        }
+
+        fclose(f);
+        free(line);
+
+        return ret;
+}
+
+static void add_if_not_empty(struct json_object *obj, const char *key, struct json_object *val) {
+        if (json_object_array_length(val))
+                json_object_object_add(obj, key, val);
+        else
+                json_object_put(val);
+}
+
+static bool interface_file_exists(const char *ifname, const char *name) {
+        const char *format = "/sys/class/net/%s/%s";
+        char path[strlen(format) + strlen(ifname) + strlen(name)];
+        snprintf(path, sizeof(path), format, ifname, name);
+
+        struct stat s;
+        return !stat(path, &s);
```

Comment:
Why not use `!access(path, F_OK)` instead of `!stat(…)`? That wouldn't require an additional temporary struct.